Adds table flipping! Take 2.0

Singul0 opened the pull request:
## About The Pull Request
Revives the long dead #80348. Right click a table to flip it over, Useful for makeshift cover or ragequitting. Their integrity is equal to that of the table you flipped. Some tables aren't able to be flipped (reinforced tables, roller tables, etc).

I refactored it to be less snowflakey. so hopefully this passes with only minor changes
![290775902-48414bb3-aaa9-467d-8edf-a170a98b1506](https://github.com/user-attachments/assets/3627c537-043e-4829-b38b-f68a7d382167)
![gambar](https://github.com/user-attachments/assets/ce04a6d1-e480-46f9-8913-041f86db8a9a)
## Why It's Good For The Game
I felt like we're lacking in the makeshift defense items. This is a quicky way to plop up a shitty barricade for you to guard with. and also really cool for roleplaying and such. also, this categorically goes hard:
![gambar](https://github.com/user-attachments/assets/af593068-d9f3-49b0-9102-989ce2b4d3fb)
## Changelog
:cl:
add: You can now flip tables by right clicking them!
/:cl:

Singul0 committed:
V0.5

Singul0 committed:
still wip

Singul0 committed:
fixes projectile bug

Singul0 committed:
almost done, just icons change now

Singul0 committed:
icons done

Singul0 committed:
finishing touches

Singul0 committed:
Merge branch 'master' into table-flip-2

Singul0 committed:
this is fine, trust me

Singul0 committed:
aaaaa

Dawnseer commented:
Can you set it back up after the shooting is over and drinking can continue?

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -75,6 +86,43 @@
 
+//proc that adds elements present in normal tables
+/obj/structure/table/proc/unflip_table()
+	make_climbable()
+	AddElement(/datum/element/give_turf_traits, turf_traits)
+	AddElement(/datum/element/footstep_override, priority = STEP_SOUND_TABLE_PRIORITY)
```
For the record I really don't vibe with this, not sure if there's a better way though.

Singul0 replied:
it pains me too

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -100,11 +148,44 @@
+	if(istype(leaving, /obj/projectile))
+		return
+
+	if(direction == dir)
+		return COMPONENT_ATOM_BLOCK_EXIT
```
```suggestion
	if(direction == dir)
		leaving.Bump(src)
		return COMPONENT_ATOM_BLOCK_EXIT
```

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -100,11 +148,44 @@
 
+/obj/structure/table/proc/on_exit(datum/source, atom/movable/leaving, direction)
+	SIGNAL_HANDLER
+	if(!is_flipped)
+		return
```
```suggestion
	if(!is_flipped)
		return
		
	if(leaving.movement_type & PHASING)
		return

	if(leaving == src)
		return
```
Otherwise will result in immovable tables

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -100,11 +148,44 @@
+	SIGNAL_HANDLER
+	if(!is_flipped)
+		return
+	if(type == /obj/structure/table/glass) //Glass table, jolly ranchers pass
+		if(istype(leaving) && (leaving.pass_flags & PASSGLASS))
+			return
```
A) Should be istype
B) Would prefer if you unsnowflaked this and made this a param on tables instead

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -100,11 +148,44 @@
+	. = ..()
+	if(!is_flipped)
+		return
+	if(type == /obj/structure/table/glass) //Glass table, jolly ranchers pass
+		if(istype(mover) && (mover.pass_flags & PASSGLASS))
+			return TRUE
```
Same as above, should probably be a value on tables

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -100,11 +148,44 @@
+	if(direction == dir)
+		return COMPONENT_ATOM_BLOCK_EXIT
+
+/obj/structure/table/CanAllowThrough(atom/movable/mover, border_dir)
+	. = ..()
+	if(!is_flipped)
+		return
```
```suggestion
/obj/structure/table/CanAllowThrough(atom/movable/mover, border_dir)
	. = ..()
	if(.)
		return
		
	if(!is_flipped)
		return FALSE
		
```

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -154,6 +237,56 @@
+			layer = ABOVE_MOB_LAYER
+
+		is_flipped = TRUE
+		flip_table()
```
Early return here

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -154,6 +237,56 @@
+		//change icons
+		layer = LOW_ITEM_LAYER
+		var/new_dir = get_dir(user, src)
+		if(new_dir == NORTHEAST || new_dir == SOUTHEAST) // Dirs need to be part of the 4 main cardinal directions so proc/CanAllowThrough isn't fucky wucky
+			new_dir = EAST
+		if(new_dir == NORTHWEST || new_dir == SOUTHWEST)
+			new_dir = WEST
```
```suggestion
		if(new_dir & EAST) // Dirs need to be part of the 4 main cardinal directions so proc/CanAllowThrough isn't fucky wucky
			new_dir = EAST
		else if(new_dir & WEST)
			new_dir = WEST
```

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -154,6 +237,56 @@
+			return
+
+		//change icons
+		layer = LOW_ITEM_LAYER
```
Any reason all of this logic isn't in flip_table()?

SmArtKar commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -154,6 +237,56 @@
+		return FALSE
+	if(movable_entity.throwing || (movable_entity.movement_type & (FLOATING|FLYING))) //Thing isn't flying/floating
+		return FALSE
+
```
Check for TRAIT_MOB_ELEVATED on mobs please, so it works with TRAIT_IGNORE_ELEVATION (like Melbert's antigravity PR)

Jacquerel replied:
I was also going to suggest this until I noticed that you did; this means that you can flip tables with xeno larva underneath to reveal them which is very soulful.

SmArtKar reviewed (changes requested):
Bronze and abductor tables shouldn't be flippable, they're made from plasteel equivalent materials and look weird when flipped. (Kinda odd that they're not r-table subtypes, I should probably make them those in a separate PR)

The PR itself is amazing, obviously a lot of code comments but I would be very happy to see this in the codebase

Singul0 commented:
> will the items fall from table to neighbour tile when you flip it? 😀

Yep

Jacquerel commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -154,6 +237,56 @@
+		return FALSE
+	if (movable_entity.anchored) //Thing isn't anchored
+		return FALSE
+	if (movable_entity.invisibility > SEE_INVISIBLE_LIVING) //Thing isnt incorporeal ie. ghosts/revenants
+		return FALSE
+	if(!ismob(movable_entity) && !isobj(movable_entity)) //Thing isn't an obj or mob
```
```suggestion
	if(!isliving(movable_entity) && !isobj(movable_entity)) //Thing isn't an obj or mob
```
Instead of filtering out observers via invisibility why not filter them out via type?

timothymtorres commented:
What happens if a mob is standing on top of it? (IMO you shouldn't be allowed to flip it in that case)

Kiromaka commented:
> What happens if a mob is standing on top of it? (IMO you shouldn't be allowed to flip it in that case)

It should fall on its head and die, instant 300 brute to brain.

MrMelbert commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -100,11 +148,44 @@
 
 /obj/structure/table/examine(mob/user)
 	. = ..()
+	if(is_flipped)
+		. += span_notice("It's been flipped on its side!")
```
If you ask me this is kind of redundant. I mean you can see it's flipped

Singul0 replied:
figured that some more indicators would be nice, since if it's your first time seeing this you'd be wondering what's happenning

MrMelbert replied:
Idk, the sprite literally has the table on its side

Singul0 replied:
well, you know how there is a large overlap between smartest bear and the stupidest man? taking no chances

Singul0 committed:
smarkar code review

Co-authored-by: SmArtKar <44720187+SmArtKar@users.noreply.github.com>
Co-authored-by: Jacquerel <hnevard@gmail.com>

Singul0 committed:
the rest of the code review barring one

Singul0 committed:
right-o

Singul0 commented:
> My concern with this is, again, anyone adding a table would need to add a flipped icon
> 
> Any way you can instead edit the transform matrix of the table?
![gambar](https://github.com/user-attachments/assets/e59ddd39-96ec-4914-b2fb-7eef718e5d49)

Doesn't look too bad much better than I anticipated! the only sprite that falters is the fancy carpeted ones



> What happens if a mob is standing on top of it? (IMO you shouldn't be allowed to flip it in that case)

They get thrown over to the adjacent dir the table is flipped to, along with any other items, I think it's pretty soulful. a tidie tries to get over the counter and you just flip the table off

Singul0 committed:
fixes throwing, TODO; Matrices

Singul0 committed:
guh

Singul0 committed:
minor fixes, apparently the 7 years old smooth_icon var hack is defunct? or atleast I dont notice anything different from changing it

Singul0 committed:
adds spawners

optimumtact commented:
aren't there a lot of map areas relying on tables to block/slow progress into areas?

SmArtKar commented:
> aren't there a lot of map areas relying on tables to block/slow progress into areas?

Most secure areas use reinforced tables which are unflippable

Singul0 committed:
it's quite shrimple really, adds matrices

I FUCKING HATE BYOND MATRICES AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

Singul0 committed:
removes make_climbable proc

Singul0 commented:
doneso

Singul0 committed:
Revert "removes make_climbable proc"

This reverts commit 4c5fe3544baecc68d988987c5927f50d1df798c9.

Wallemations commented:
> Doesn't look too bad much better than I anticipated! the only sprite that falters is the fancy carpeted ones. maybe make the uh, matrix standard with an override if you want to?

I think the matrix shouldn't be standard since you already have a lot of the flipped sprites created beforehand. It should be the fallback in case something doesn't have a flipped sprite. Code-wise that would pretty much be the same that you said tbh

timothymtorres commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -154,6 +292,41 @@
+		return FALSE
+	if (movable_entity.anchored) //Thing isn't anchored
+		return FALSE
+	if(!isliving(movable_entity) && !isobj(movable_entity)) //Thing isn't an obj or mob
+		return FALSE
```
Since you are not supporting throwing living mobs off the table, you should restrict table flipping in `attack_hand_secondary` if a mob is standing on it.

Singul0 committed:
matrices is the override instead

Singul0 commented:
alrighty, it uses the pre-baked flipped table sprites as a default now, with matrices as a fallback option if the tables don't have a flipped table icon

Singul0 committed:
that's it, I think

Jacquerel commented on code/game/objects/structures/tables_racks.dm:
```diff
@@ -37,6 +40,16 @@
+	/// Whether or not when flipped, it ignores PASS_GLASS flag
+	var/is_transparent = FALSE
+	/// If you don't have sprites for flipped tables, you can use matrices instead. looks ever-slightly worse.
+	var/use_matrices_instead = FALSE
+	/// Matrix to return to on unflipping table
+	var/matrix/original_matrix
```
Quick nitpick before we merge, please mention this regards being flipped in the var name

Singul0 committed:
last nitpick

Singul0 commented:
doneso

Singul0 commented:
> So erm, I'm porting this to the downstream that this was originally ported from and ironically this breaks horribly if you try to put one of your `/obj/effect/spawner/structure/flipped_table` 's on a map.
> 
> ![firefox_v2JOP29sr2](https://private-user-images.githubusercontent.com/13398309/431362901-07b41094-7e49-4762-b889-e7892a4dddcb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQxMTk1MTQsIm5iZiI6MTc0NDExOTIxNCwicGF0aCI6Ii8xMzM5ODMwOS80MzEzNjI5MDEtMDdiNDEwOTQtN2U0OS00NzYyLWI4ODktZTc4OTJhNGRkZGNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA4VDEzMzMzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIxMmNjYWJjODcwOTI1ZDBhNGY4MTdhYWE1ODA1YWVmZWQxODg1MDU3YmIwOGIzNDc3YjRjOTE0NjRkYWU2N2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.y1XfHdn581vAakEx6NHr7PeMFQ8S4yO68OONtjFfM5c)
> 
> TG doesn't have any placed in maps so perhaps it just slipped by you but it is causing us a bit of a headache

Ah shucks, fault's on me for adding a debugging tool as a feature last minute. It's probably something to do with async-ed smooth procs still running while we cop out the smooth_icons. one sec i'll review it with a fix soon ish
